presto-parser/src/main/java/io/prestosql/sql/tree/Query.java

```diff
@@ -29,43 +29,49 @@
     private final QueryBody queryBody;
     private final Optional<OrderBy> orderBy;
     private final Optional<String> limit;
+    private final Optional<String> fetch;
```

Comment:
This approach makes it theoretically possible to construct syntactically illegal queries (i.e., with both LIMIT and FETCH FIRST clauses, potentially conflicting). A better approach is to replace the `Optional<String> limit` with an `Optional<Node> limit` and fill it in with either a `Limit` or a `FetchFirst` AST node (both of which need to be introduced). 

This would also make the analysis more regular -- it just needs to delegate to method to analyze the corresponding node type (there'd be two of them), each of which would be responsible for ensuring the values are legal, etc., and record the inferred limit in the Analysis object to be used later by the planner.

Reply:
Please notice that a query mixing LIMIT and FETCH FIRST clauses wouldn't pass through the parser. Despite that I agree that they should share a field in `Query` / `QuerySpecification`, as they are mutually exclusive.
Should I introduce a common superclass for `Limit` and `FetchFirst` nodes? Or maybe for future `Offset` node too?

Reply:
> Should I introduce a common superclass for Limit and FetchFirst

If we can come up with a reasonable name for it, yes -- there may not be a good generic concept that describes both. Otherwise, it's fine to keep as Node for now and maybe do a runtime check to ensure it's one of the supported types.